Classes/common/CDTIncrementalStore.m

```diff
@@ -0,0 +1,2398 @@
+ *  a docId/ref.
+ *
+ *  @param docID   docID
+ *  @param context context
+ *  @param version version
+ *  @param error   error
+ *
+ *  @return dictionary or nil with error
+ */
+- (NSDictionary *)valuesFromDocID:(NSString *)docID
+                      withContext:(NSManagedObjectContext *)context
+                       versionPtr:(uint64_t *)version
+                            error:(NSError **)error
+{
+    NSError *err = nil;
+
+    CDTDocumentRevision *rev = [self.datastore getDocumentWithId:docID error:&err];
+    if (!rev) {
+        if (error) *error = err;
+        return nil;
+    }
+    NSMutableDictionary *values = [NSMutableDictionary dictionary];
+    for (NSString *name in rev.body) {
+        if ([name isEqualToString:CDTISObjectVersionKey]) {
+            *version = [rev.body[name] longLongValue];
+            continue;
+        }
+        if ([name hasPrefix:CDTISPrefix]) {
+            continue;
+        }
+
+        id value = [self decodeProperty:name
+                                fromDoc:rev.body
+                          withBlobStore:rev.attachments
+                            withContext:context];
+        if (!value) {
+            // Dictionaries do not take nil, but Values can't have NSNull.
+            // Apparently we just skip it and the properties faults take care
+            // of it
+            continue;
+        }
+        values[name] = value;
+    }
+
+    return [NSDictionary dictionaryWithDictionary:values];
+}
+
+#pragma mark - Push/Pull with Remote methods
+
+- (BOOL)commWithRemote:(CDTReplicator *)rep
+                 error:(NSError **)error
+          withProgress:(CDTISProgressBlock)progress
+{
+    NSError *err = nil;
+
+    CDTReplicatorState state = rep.state;
+    // we can only start from pending
+    if (state != CDTReplicatorStatePending) {
+        NSString *stateName = nil;
+        switch (state) {
+            case CDTReplicatorStatePending:
+                break;
+            case CDTReplicatorStateComplete:
+                stateName = @"CDTReplicatorStateComplete";
+                break;
+            case CDTReplicatorStateError:
+                stateName = @"CDTReplicatorStateError";
+                break;
+            case CDTReplicatorStateStopped:
+                stateName = @"CDTReplicatorStateStopped";
+                break;
+            default:
+                stateName = [NSString stringWithFormat:@"Unknown replicator state: %@", @(state)];
+                break;
+        }
+        if (error) {
+            NSString *s =
+                [NSString localizedStringWithFormat:@"Replicator in state: %@", stateName];
+
+            *error = [NSError errorWithDomain:CDTISErrorDomain
+                                         code:CDTISErrorSyncBusy
+                                     userInfo:@{NSLocalizedFailureReasonErrorKey : s}];
+        }
+        return NO;
+    }
+
+    if (self.progressBlock) {
```

Comment:
@mkistler says:
> Suggest a GCD/NSOperation solution.  This may enable some sync/conflict mechanism to be provided once the `optLock` stuff is in place.